galsim/config/input.py

```diff
@@ -294,6 +298,50 @@ def GetInputObj(input_type, config, base, param_name):
     return base['input_objs'][input_type][num]
 
 
+valid_input_types = {}
+
+def RegisterInputType(input_type, init_func, types, kwargs_func=None, has_nobj=False, 
+                      file_scope=False, setup_func=None):
+    """Register an input type for use by the config apparatus.
+
+    @param input_type       The name of the type in config['input']
+    @param init_func        A function or class name to use to build the input object.
+    @param types            A list of value or object types that use this input type.
+                            These items will have their "current" values invalidated when the
+                            input object changes.
+    @param kwargs_func      A function to get the initialization kwargs if the regular
+                            _req, _opt, etc. kind of initialization will not work. The call
+                            signature is:
+                                kwargs, safe = GetKwargs(config, base)
+                            [default: None, which means use the regular initialization]
+    @param has_nobj         Whether the object can be used to automatically determine the number
+                            of objects to build for a given file or image.  If True, it must have
+                            a getNObjects() method and also a construction kwargs _nobjects_only
+                            to efficiently do only enough to calculate nobjects. [default: False]
+    @param file_scope       Whether the class might be relevant at the file- or image-scope level,
+                            rather than just at the object level.  Notably, this is true for dict.
+                            [default: False]
+    @param setup_func       A function to call at the start of each image. The call signature is
+                                Setup(input_obj, config, base, logger)
+                            [default: None]
+    """
+    valid_input_types[input_type] = {
+        'init' : init_func,
+        'types' : types,
+        'kwargs' : kwargs_func,
+        'nobj' : has_nobj,
+        'file' : file_scope,
+        'setup' : setup_func
+    }
```

Comment:
Again, I wonder if a class hierarchy would make it easier for extenders to understand and implement the interface they need to support.